debcraft/models/project.py

```diff
@@ -16,12 +16,84 @@
 
 """Package metadata for Debcraft."""
 
+import enum
+import re
+from typing import Annotated
+
+import pydantic
 from craft_application import models
+from typing_extensions import Self
+
+from debcraft.models.package import Package
+
+DEBIAN_PACKAGE_NAME_REGEX = r"^[a-z0-9][a-z0-9.+-]+$"
+"""A regular expression to implement Debian package name rules.
+
+Based on the rules provided at:
+https://www.debian.org/doc/debian-policy/ch-controlfields.html#s-f-source
+"""
+
+
+def _validate_debian_package_name(name: str) -> str:
+    """Validate a Debian package name."""
+    if not re.fullmatch(DEBIAN_PACKAGE_NAME_REGEX, name):
+        raise ValueError(
+            "package names must consist only of lower case letters (a-z), digits "
+            "(0-9), plus (+) and minus (-) signs, and periods (.). They must be at "
+            "least two characters long and must start with an alphanumeric character."
+        )
+    return name
+
+
+DebianPackageName = Annotated[
+    str,
+    pydantic.Field(min_length=2, pattern=DEBIAN_PACKAGE_NAME_REGEX),
+    pydantic.BeforeValidator(_validate_debian_package_name),
```

Comment:
how does `pattern=DEBIAN_PACKAGE_NAME_REGEX` interact with this validator, that *also* matches against the regex?